admin/yum.py

```diff
@@ -0,0 +1,175 @@
+    """
+    # TODO avoid circular imports when importing these somewhere else
+    # Reviewer - before I do, this might be easier if the logic to get
+    # file names were in release.py, is that more sensible?
+    from release import make_rpm_version
+    from admin.packaging import (ARCH_REQUIREMENT, Distribution,
+        package_filename)
+
+    rpm_version = make_rpm_version(intent.flocker_version)
+    distribution = Distribution(
+        name=intent.distro_name,
+        version=intent.distro_version,
+    )
+    package_type = distribution.package_type()
+    s = requests.Session()
+    # Tests use a local package repository
+    s.mount('file://', FileAdapter())
+
+    downloaded_packages = set()
+    for package in intent.packages:
+        package_name = package_filename(
+            package_type=package_type,
+            package=package,
+            architecture=ARCH_REQUIREMENT[package],
+            rpm_version=rpm_version,
+        )
+        url = intent.source_repo + '/' + package_name
+        local_path = intent.target_path.child(package_name).path
+        download = s.get(url)
+        download.raise_for_status()
+        content = download.content
+        with open(local_path, "wb") as local_file:
+            local_file.write(content)
+        downloaded_packages.add(package_name)
+
+    return downloaded_packages
+
+
+@attributes([
+    "repository_path",
+    "existing_metadata",
+])
+class CreateRepo(object):
+    """
+    Create repository metadata, and return filenames of new and changed
+    metadata files.
+
+    Note that this returns a list with the prefixes stripped.
+
+    :ivar FilePath repository_path: Location of rpm files to create a
+        repository from.
+    :ivar set existing_metadata: Filenames of existing metadata files.
+    """
+
+
+@sync_performer
+def perform_create_repository(dispatcher, intent):
+    """
+    See :class:`CreateRepo`.
+
+    :return: List of new and modified rpm metadata filenames.
+    """
+    # The update option means that this is faster when there is existing
+    # metadata but has output starting "Could not find valid repo at:" when
+    # there is not existing valid metadata.
+    check_call([
+        b'createrepo',
+        b'--update',
+        b'--quiet',
+        intent.repository_path.path])
+    return _list_new_metadata(
+        repository_path=intent.repository_path,
+        existing_metadata=intent.existing_metadata)
+
+
+def _list_new_metadata(repository_path, existing_metadata):
+    """
+    List the filenames of new and changed repository metadata files.
+
+    :param FilePath repository_path: Location of repository to list repository
+        metadata from.
+    :param set existing_metadata: Filenames of existing metadata files.
+    """
+    all_metadata = set([os.path.basename(path.path) for path in
+                        repository_path.child('repodata').walk()])
+    new_metadata = all_metadata - existing_metadata
```

Comment:
Since
- we don't download old metadata
- metadata files are small
- we are unlikely to generate exactly the same metadata twice

we probably don't need to bother excluding existing metadata here.
If we were going to do anything like this, it would probably make sense to list pre-existing *local* metadata, and exclude that (this could be done as part of this effect).